### ISSUE TITLE
(A11y severity 2) Document preview control accessibility issues

### ISSUE THREAD

**Issue by sathomas** [oaeproject/3akai-ux#3927]
- Even though a keyboard or screen reader user may select a specific page to read or navigate to using the page number input, this simply scrolls the viewer area. Once the user navigates into the frame they begin reading and navigating on the first page, even though this page may not be visible.
- Navigating through the preview document can be very difficult for users since there are no focus indicators and the page scroll doesn’t appear to match wherever the link is found.

**Comment by sathomas** [comment]
The document lives in a scrolling container (`#documentpreview-content`). By clicking the buttons or enter text in the page number `<input>`, the code adjusts the scroll position to bring the relevant page to the top.

Screen readers, however, don't consider scroll position. So, for example, if a user moves to page 100 using the toolbar and _then_ navigates to the content, the screen reader starts reading from the first page, not from page 100.

I'm researching approaches that might solve this problem.

**Comment by sathomas** [comment]
It doesn't look like there's a simple way to address this issue, but I have an idea that might not be too terrible.

* Hide the existing toolbar from screen readers. It's pretty much useless to them anyway. Zooming and full-screen aren't helpful, and the pagination controls (as noted by this issue) aren't effective.
* Create a secondary "toolbar" that's visible _only_ to screen readers. This secondary toolbar would implement the standard pagination controls common on the web (though with all the right accessibility attributes). E.g. the Bootstrap example with the addition of "first page" and "last page" buttons (though, of course, it wouldn't be _visible_ on the page).

![screen shot 2015-01-15 at 1 00 09 pm](https://cloud.githubusercontent.com/assets/1731910/5765795/5ca5d378-9cca-11e4-84b4-c913824fc8c9.png)

* Add `id` attributes to all of the pages in the document, and link the sr-only toolbar to those pages with fragment URLs  (`<a href="#page-10">10</a>`)

Some implications of the approach:

* Since we're lazy-loading the pages, we'll have to use JavaScript to intercept clicks on the sr-only toolbar, load the page, and then set `window.location.href` appropriately. Should be simple enough.
* Our content URLs will now include fragments (`https://oae.gatech.edu/content/gatech/Q1MThz3Z#page-10`). I don't know if this will break anything in the backend, but if it works, I actually think it's a usability improvement in general. In particular

    * If a user bookmarks or shares (e.g. via email) a content link, that link would open directly to the relevant page. (Of course, that will take a small bit of code in the front-end to jump to that page; pretty easy, though.)
    * For full-text search, we could link directly to the (first) page in the document with the search term. I think this is a significant gain. As it stands, if full-text search points to a document, there's practically no way to find where in the document. Because all the content is lazy-loaded, and because pdf2htmlEX uses crazy fonts, searching on the page (⌘-F) doesn't work. The only option is to read through the document. At least with this approach we could start on the right page instead of from the very beginning.

Details still under consideration:

* I'm not completely sure of the best way to handle large documents (e.g. 100 pages). Even though the sr-only toolbar isn't visible, I'm pretty sure it wouldn't be helpful to have a navbar with 100 `<li>` elements. It would be nice to avoid the complexity of a dynamic toolbar, but I'm not sure if that's possible.
* Related to the previous bullet, we should provide a way for screen reader users to jump directly to a specific page. Probably the best way is just to replicate the page number `<input>` from the visible toolbar on the sr-only toolbar, but there might be a better way that integrates a little more smoothly with the pagination.

Over to @nicolaasmatthijs, @simong, and @stuartf for thoughts.

**Comment by sathomas** [comment]
Did some testing and screen readers (well, VoiceOver) don't handle URL fragments the way I was expecting. So the approach above won't solve the issue, at least without more work. Any thoughts on the page fragments in our URLs would still be welcome, as other possible solutions may well benefit from them.

**Comment by nicolaasmatthijs** [comment]
@sathomas : Is there anything we can do here with adding a `tabIndex` to the currently active page and maybe setting focus appropriately?

> Any thoughts on the page fragments in our URLs would still be welcome, as other possible solutions may well benefit from them.

I like the idea of recording the current page number in the URL. However, we'd have to make sure that clicking the browser back button doesn't just take you to the previous page of the document. For a long document, that would be too intrusive. When scrolling through a document and then sending the URL to someone, the expected behaviour might not always be that that user is taken to the same page straight away either, so we probably want to be careful and ensure that we don't make the behaviour worse.

**Comment by sathomas** [comment]
> Is there anything we can do here with adding a tabIndex to the currently active page and maybe setting focus appropriately?

That's pretty much exactly what I've done in PR https://github.com/oaeproject/3akai-ux/pull/3911. It works surprisingly well. There's a separate question re: performance that's holding up that PR for now.

**Comment by nicolaasmatthijs** [comment]
@sathomas : If we go with the approach in #3911, is there any outstanding work for this ticket?

**Comment by sathomas** [comment]
@nicolaasmatthijs #3911 helps a lot, but there's still a problem accessing deep pages (e.g. page number 50 in a long document). Sighted users can use the toolbar to enter a page number directly and the preview will scroll to that page. I haven't (yet) completely figured out how to make that happen for screen readers. I have some ideas, but need to experiment further.

**Comment by sathomas** [comment]
> However, we'd have to make sure that clicking the browser back button doesn't just take you to the previous page of the document. For a long document, that would be too intrusive.

I agree (and it's easy enough to use `history.replaceState()` to update the URL without creating a new entry in the browser history). So we could add page number fragments to the document for bookmarks, sharing, and search results (once the backend is modified to support them).

Unfortunately, if we don't update the browser history, I haven't (yet) found a way to make screen readers advance to the page. Still working on that, though.
